File: php/class/auction.php
<?php
$auctionJsonFieldMapping = array(
  // API level
  "status" => "s",
  "data" => "d",
  "message" => "m",
  "key" => "k",
  "requestStart" => "t",
  "elapsed" => "e",
  // Auction
  "num" => "n",
  "startTime" => "st",
  "collectionDeadline" => "cd",
  "location" => "l",
  "auctionPdf" => "ap",
  "resultPdf" => "rp",
  "itemPdfList" => "ipl",
  "remarks" => "r",
  "lotList" => "ll",
  "lotCount" => "lc",
  "transactionTotal" => "tt",
  "auctionStatus" => "as",
  // "status" => "s",
  "lastUpdate" => "lu",
  // ------
  // AuctionLot
  "type" => "t",
  "lotNum" => "ln",
  "gldFileRef" => "gr",
  "reference" => "rf",
  "department" => "dp",
  "contact" => "co",
  "contactNumber" => "cn",
  "contactLocation" => "cl",
  // "remarks" => "r",
  "itemCondition" => "ic",
  "featured" => "f",
  "icon" => "i",
  "photoUrl" => "pu",
  "photoReal" => "pr",
  "photoAuthor" => "pa",
  "photoAuthorUrl" => "pau",
  "descriptionEn" => "den",
  "descriptionTc" => "dtc",
  "descriptionSc" => "dsc",
  "itemList" => "il",
  "specialInspection" => "si",
  "inspectionDateList" => "idl",
  "tranCurrency" => "tc",
  "tranPrice" => "tp",
  "tranStatus" => "ts",
  // "status" => "s",
  // "lastUpdate" => "lu",
  // ------
  // Auction Item
  // "icon" => "i",
  "description" => "d",
  "quantity" => "q",
  "unit" => "u",
  // ------
  // Inspection Date
  "dayOfWeek" => "dow",
  //"startTime" => "st",
  "endTime" => "et",
  // ------
  // Auction Search
  "auctionId" => "aid",
  "lotId" => "lid",
  // ------
  // Auction Lot Grid Item
  "auctionNum" => "an",
);

class Auction implements JsonSerializable {
  private $id;
  private $num;
  private $startTime;
  private $collectionDeadline;
  private $location;
  private $auctionPdf;
  private $resultPdf;
  private $remarks;
  private $lotCount;
  private $transactionTotal;
  private $itemPdfList;
  private $lotList;
  private $auctionStatus;
  private $status;
  private $lastUpdate;

  public function __construct($id, $num, $startTime, $collectionDeadline, $location, $auctionPdf, $resultPdf, $remarks, $lotCount, $transactionTotal, $auctionStatus, $status, $lastUpdate) {
    $this->id = $id;
    $this->num = $num;
    $this->startTime = $startTime;
    $this->collectionDeadline = $collectionDeadline;
    $this->location = $location;
    $this->auctionPdf = $auctionPdf;
    $this->resultPdf = $resultPdf;
    $this->remarks = $remarks;
    $this->lotCount = $lotCount;
    $this->transactionTotal = $transactionTotal;
    $this->itemPdfList = array();
    $this->lotList = array();
    $this->auctionStatus = $auctionStatus;
    $this->status = $status;
    $this->lastUpdate = $lastUpdate;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}

class AuctionLot implements JsonSerializable {
  private $id;
  private $type;
  private $lotNum;
  private $gldFileRef;
  private $reference;
  private $department;
  private $contact;
  private $contactNumber;
  private $contactLocation;
  private $remarks;
  private $itemCondition;
  private $descriptionEn;
  private $descriptionTc;
  private $descriptionSc;

  private $featured;
  private $icon;
  private $photoUrl;
  private $photoReal;
  private $photoAuthor;
  private $photoAuthorUrl;
  private $itemList;
  private $specialInspection;
  private $inspectionDateList;
  private $tranCurrency;
  private $tranPrice;
  private $tranStatus;
  private $status;
  private $lastUpdate;

  public function __construct($id, $type, $lotNum, $gldFileRef, $reference, $department, $contact, $number, $location, $remarks, $itemCondition, $featured, $icon, 
                              $photoUrl, $photoReal, $photoAuthor, $photoAuthorUrl, $descriptionEn, $descriptionTc, $descriptionSc, $tranCurrency, $tranPrice, $tranStatus, $status, $lastUpdate) {
    $this->id = $id;
    $this->type = $type;
    $this->lotNum = $lotNum;
    $this->gldFileRef = $gldFileRef;
    $this->reference = $reference;
    $this->department = $department;
    $this->contact = $contact;
    $this->contactNumber = $number;
    $this->contactLocation = $location;
    $this->remarks = $remarks;
    $this->itemCondition = $itemCondition;

    $this->featured = $featured;
    $this->icon = $icon;
    $this->photoUrl = $photoUrl;
    $this->photoReal = $photoReal;
    $this->photoAuthor = $photoAuthor;
    $this->photoAuthorUrl = $photoAuthorUrl;
    $this->descriptionEn = $descriptionEn;
    $this->descriptionTc = $descriptionTc;
    $this->descriptionSc = $descriptionSc;

    $this->itemList = array();
    $this->specialInspection = false;
    $this->inspectionDateList = array();
    $this->tranCurrency = $tranCurrency;
    $this->tranPrice = $tranPrice;
    $this->tranStatus = $tranStatus;
    $this->status = $status;
    $this->lastUpdate = $lastUpdate;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}

class AuctionItem implements JsonSerializable {
  private $id;
  private $icon;
  private $description;
  private $quantity;
  private $unit;

  public function __construct($id, $icon, $description, $quantity, $unit) {
    $this->id = $id;
    $this->icon = $icon;
    $this->description = $description;
    $this->quantity = $quantity;
    $this->unit = $unit;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}

class AuctionLotSearch implements JsonSerializable {
  private $auctionId;
  private $auctionNum;
  private $startTime;
  private $auctionStatus;
  private $lotId;
  private $lotNum;
  private $type;
  private $description;
  private $featured;
  private $icon;
  private $photoUrl;
  private $photoReal;
  private $photoAuthor;
  private $photoAuthorUrl;
  private $tranCurrency;
  private $tranPrice;
  private $tranStatus;

  public function __construct($auction_id, $auction_num, $start_time, $auction_status, $lot_id, $lot_num, $type, $description, $featured, $icon, 
                              $photoUrl, $photoReal, $photoAuthor, $photoAuthorUrl, $tranCurrency, $tranPrice, $tranStatus) {
    $this->auctionId = $auction_id;
    $this->auctionNum = $auction_num;
    $this->startTime = $start_time;
    $this->auctionStatus = $auction_status;
    $this->lotId = $lot_id;
    $this->lotNum = $lot_num;
    $this->type = $type;
    $this->description = $description;
    $this->featured = $featured;
    $this->icon = $icon;

    $this->photoUrl = $photoUrl;
    $this->photoReal = $photoReal;
    $this->photoAuthor = $photoAuthor;
    $this->photoAuthorUrl = $photoAuthorUrl;
    $this->tranCurrency = $tranCurrency;
    $this->tranPrice = $tranPrice;
    $this->tranStatus = $tranStatus;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}

class AuctionItemSearch implements JsonSerializable {
  private $auctionId;
  private $startTime;
  private $auctionStatus;
  private $lotId;
  private $type;
  private $featured;
  private $photoUrl;
  private $photoReal;
  private $photoAuthor;
  private $photoAuthorUrl;
  private $tranCurrency;
  private $tranPrice;
  private $tranStatus;
  private $icon;
  private $description;
  private $quantity;
  private $unit;

  public function __construct($auction_id, $start_time, $auction_status, $lot_id, $type, $featured, $photoUrl, $photoReal, $photoAuthor, $photoAuthorUrl, $tranCurrency, $tranPrice, $tranStatus, $icon, $description, $quantity, $unit) {
    $this->auctionId = $auction_id;
    $this->startTime = $start_time;
    $this->auctionStatus = $auction_status;
    $this->lotId = $lot_id;
    $this->type = $type;
    $this->featured = $featured;
    $this->photoUrl = $photoUrl;
    $this->photoReal = $photoReal;
    $this->photoAuthor = $photoAuthor;
    $this->photoAuthorUrl = $photoAuthorUrl;
    $this->tranCurrency = $tranCurrency;
    $this->tranPrice = $tranPrice;
    $this->tranStatus = $tranStatus;
    $this->icon = $icon;
    $this->description = $description;
    $this->quantity = $quantity;
    $this->unit = $unit;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}

class InspectionDate implements JsonSerializable {
  private $dayOfWeek;
  private $startTime;
  private $endTime;

  public function __construct($day, $start_time, $end_time) {
    $this->dayOfWeek = $day;
    $this->startTime = $start_time;
    $this->endTime = $end_time;
  }

  public function __get($property) {
    if (property_exists($this, $property)) {
      return $this->$property;
    }
  }

  public function __set($property, $value) {
    if (property_exists($this, $property)) {
        $this->$property = $value;
    }

    return $this;
  }

  public function jsonSerialize() {
    $vars = get_object_vars($this);
    return $vars;
  }
}
?>

File: lib/class/auction.dart
import 'package:intl/intl.dart';
// import 'package:logger/logger.dart';
import '../includes/enums.dart';
import '../includes/revelation.dart' as revelation;

class Auction {
  Auction(this.id, this.auctionNum, this.startTime, this.collectionDeadline, this.location, this.auctionPdfUrl, this.resultPdfUrl, this.itemPdfList,
      this.remarks, this.lotCount, this.transactionTotal, this.lotList, this.status, this.lastUpdate);

  factory Auction.empty() {
    return Auction(0, '', DateTime(1900), DateTime(1900), '', '', '', <AuctionItemPdf>[], '', 0, 0, <AuctionLot>[], AuctionStatus.None, DateTime(1900));
  }

  factory Auction.fromJson(Map<String, dynamic> json) {
    final List<dynamic> ipl = json['ipl'] as List<dynamic>;

    return Auction(
      json['id'] as int,
      json['n'] as String,
      DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['st'] as String),
      DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['cd'] as String),
      json['l'] as String,
      json['ap'] as String,
      json['rp'] as String,
      ipl.map((dynamic ip) {
        final Map<String, String> itemPdf = ip as Map<String, String>;

        if (itemPdf['t'] != null && itemPdf['url'] != null) {
          return AuctionItemPdf(itemPdf['t']!, itemPdf['url']!);
        } else {
          return AuctionItemPdf('', '');
        }
      }).toList()
        ..removeWhere((AuctionItemPdf a) => a.itemType == ''),
      json['r'] as String,
      revelation.revealAuctionLotCount(json['lc'] as int, json['id'] as int),
      revelation.revealAuctionTransactionTotal(json['tt'] as int, json['id'] as int),
      <AuctionLot>[],
      getAuctionStatus(json['as'] as String),
      DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['lu'] as String),
    );
  }

  final int id;
  final String auctionNum;
  final DateTime startTime;
  final DateTime collectionDeadline;
  final String location;
  final String auctionPdfUrl;
  final String resultPdfUrl;
  final String remarks;
  final int lotCount;
  final int transactionTotal;
  final AuctionStatus status;
  final DateTime lastUpdate;
  final List<AuctionItemPdf> itemPdfList;
  final List<AuctionLot> lotList;
}

class AuctionLot {
  AuctionLot(
      this.id,
      this.auctionDate,
      this.itemType,
      this.lotNum,
      this.gldFileRef,
      this.reference,
      this.department,
      this.contact,
      this.contactNumber,
      this.contactLocation,
      this.remarks,
      this.itemCondition,
      this.description,
      this.descriptionEn,
      this.descriptionTc,
      this.descriptionSc,
      this.featured,
      this.icon,
      this.photoUrl,
      this.photoReal,
      this.itemList,
      this.specialInspection,
      this.inspectionDateList,
      this.transactionCurrency,
      this.transactionPrice,
      this.transactionStatus,
      this.lastUpdate);

  factory AuctionLot.fromJson(Map<String, dynamic> json, String lang) {
    final List<dynamic> jsonItemList = json['il'] as List<dynamic>;
    final List<AuctionItem> itemList = <AuctionItem>[];
    final List<dynamic> jsonInspectionDateList = json['idl'] as List<dynamic>;
    final List<InspectionDate> inspectionDateList = <InspectionDate>[];

    for (final dynamic jsonItem in jsonItemList) {
      itemList.add(AuctionItem.fromJson(jsonItem as Map<String, dynamic>));
    }

    for (final dynamic jsonItem in jsonInspectionDateList) {
      inspectionDateList.add(InspectionDate.fromJson(jsonItem as Map<String, dynamic>));
    }

    return AuctionLot(
      json['id'] as int,
      json['ad'] == null ? DateTime(1900) : DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['ad'] as String),
      json['t'] as String,
      json['ln'] as String,
      json['gr'] as String,
      json['rf'] as String,
      json['dp'] as String,
      json['co'] as String,
      json['cn'] as String,
      json['cl'] as String,
      json['r'] as String,
      json['ic'] as String,
      lang == 'tc' ? json['dtc'] as String : (lang == 'sc' ? json['dsc'] as String : json['den'] as String),
      json['den'] as String,
      json['dtc'] as String,
      json['dsc'] as String,
      json['f'] as int == 1,
      json['i'] as String,
      json['pu'] as String,
      json['pr'] as int == 1,
      itemList,
      json['si'] as int == 1,
      inspectionDateList,
      json['tc'] as String,
      jsonToDouble(json['tp']),
      json['ts'] as String,
      DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['lu'] as String),
    );
  }

  factory AuctionLot.empty() {
    return AuctionLot(0, DateTime(1900), '', '', '', '', '', '', '', '', '', '', '', '', '', '', false, '', '', false, <AuctionItem>[], false, <InspectionDate>[], '', 0.0, '',
        DateTime(1900));
  }

  final int id;
  final DateTime auctionDate;
  final String itemType;
  final String lotNum;
  final String gldFileRef;
  final String reference;
  final String department;
  final String contact;
  final String contactNumber;
  final String contactLocation;
  final String remarks;
  final String itemCondition;
  final String description;
  final String descriptionEn;
  final String descriptionTc;
  final String descriptionSc;
  final bool featured;
  final String icon;
  final String photoUrl;
  final bool photoReal;
  final List<AuctionItem> itemList;
  final bool specialInspection;
  final List<InspectionDate> inspectionDateList;
  final String transactionCurrency;
  final double transactionPrice;
  final String transactionStatus;
  final DateTime lastUpdate;

  // use the description field now
  // String get title {
  //   if (itemList.isEmpty) {
  //     return lotNum;
  //   }

  //   return itemList.map((AuctionItem auctionItem) => auctionItem.description).toList().join(', ');
  // }

  String get itemDescriptionList {
    if (itemList.isEmpty) {
      return lotNum;
    }

    return itemList.map((AuctionItem auctionItem) => '${auctionItem.description} ${auctionItem.quantity} ${auctionItem.unit}').toList().join(', ');
  }
}

class AuctionItem {
  AuctionItem(this.id, this.icon, this.description, this.quantity, this.unit);

  factory AuctionItem.fromJson(Map<String, dynamic> json) {
    return AuctionItem(
      json['id'] as int,
      json['i'] as String,
      json['d'] as String,
      json['q'] as String,
      json['u'] as String,
    );
  }

  final int id;
  final String icon;
  final String description;
  final String quantity;
  final String unit;
}

class AuctionItemPdf {
  AuctionItemPdf(this.itemType, this.pdfUrl);

  final String itemType;
  final String pdfUrl;
}

double jsonToDouble(dynamic val) {
  if (val.runtimeType == int) {
    return (val as int).toDouble();
  }
  return val as double;
}

class RelatedAuctionLot {
  RelatedAuctionLot(
      this.auctionId,
      this.auctionNum,
      this.startTime,
      this.auctionStatus,
      this.lotId,
      this.lotNum,
      this.itemType,
      this.description,
      this.featured,
      this.icon,
      this.photoUrl,
      this.photoReal,
      this.photoAuthor,
      this.photoAuthorUrl,
      this.transactionCurrency,
      this.transactionPrice,
      this.transactionStatus);

  factory RelatedAuctionLot.fromJson(Map<String, dynamic> json) {
    return RelatedAuctionLot(
      json['aid'] as int,
      json['an'] as String,
      DateFormat('yyyy-MM-dd HH:mm:ss').parse(json['st'] as String),
      getAuctionStatus(json['as'] as String),
      json['lid'] as int,
      json['ln'] as String,
      json['t'] as String,
      json['d'] as String,
      json['f'] as int == 1,
      json['i'] as String,
      json['pu'] as String,
      json['pr'] as int == 1,
      json['pa'] as String,
      json['pau'] as String,
      json['tc'] as String,
      jsonToDouble(json['tp']),
      json['ts'] as String,
    );
  }

  final int auctionId;
  final String auctionNum;
  final DateTime startTime;
  final AuctionStatus auctionStatus;
  final int lotId;
  final String lotNum;
  final String itemType;
  final String description;
  final bool featured;
  final String icon;

  final String photoUrl;
  final bool photoReal;
  final String photoAuthor;
  final String photoAuthorUrl;
  final String transactionCurrency;
  final double transactionPrice;
  final String transactionStatus;
}

class InspectionDate {
  InspectionDate(this.dayOfWeek, this.startTime, this.endTime, this.typhoonStartTime, this.typhoonEndTime);

  factory InspectionDate.fromJson(Map<String, dynamic> json) {
    return InspectionDate(
      json['dow'] as int,
      json['st'] as String,
      json['et'] as String,
      json['tst'] as String,
      json['tet'] as String, 
    );
  }

  final int dayOfWeek;
  final String startTime;
  final String endTime;
  final String typhoonStartTime;
  final String typhoonEndTime;
}
